dotCMS/src/main/java/com/dotmarketing/portlets/cmsmaintenance/factories/CMSMaintenanceFactory.java

```diff
@@ -17,138 +18,175 @@
+                int deletedRecords = removeOldVersions(startDate);
+                // Run the drop tasks iteratively, moving forward in time
+                // DROP_OLD_ASSET_ITERATE_BY_SECONDS controls how many seconds to move forward
+                // in time for each iteration - default is to iterate by 30 days
+                startDate.add(Calendar.SECOND,
+                        Config.getIntProperty("DROP_OLD_ASSET_ITERATE_BY_SECONDS", 60 * 60 * 24 * 30));
+                // We should never go past the date the user entered
+                if (startDate.getTime().after(assetsOlderThan)) {
+                    break;
+                }
+                totalRecords += deletedRecords;
+                HibernateUtil.commitTransaction();
+            } catch (final Exception ex) {
+                Logger.error(CMSMaintenanceFactory.class,
+                        String.format("An error occurred when deleting old asset versions: %s", ex.getMessage()), ex);
+                try {
+                    HibernateUtil.rollbackTransaction();
+                } catch (final DotHibernateException e) {
+                    Logger.error(CMSMaintenanceFactory.class,
+                            String.format("An error occurred when rolling the transaction back: %s", e.getMessage()),
+                            e);
+                }
+                return -1;
+            }
+        }
+        if (totalRecords > 0) {
+            CacheLocator.getCacheAdministrator().flushAll();
+        }
+        Logger.info(CMSMaintenanceFactory.class, String.format(
+                " The Drop Old Assets Versions tool has run. A total of %d records were deleted!", totalRecords));
+        return totalRecords;
+    }
 
+    /**
+     * Looks for the oldest last modified date of versionable objects in the system.
+     * Such a date will allow the tool to start deleting from the oldest to the
+     * newest version. If no date is found, the current date will be returned.
+     *
+     * @return The oldest modification date from a versionable object.
+     */
+    private static Calendar getOldestInodeDate() {
+        Calendar startDate = Calendar.getInstance();
+        final DotConnect dc = new DotConnect();
+        dc.setSQL(
+                "SELECT MIN(idate) AS min_date FROM inode WHERE type IN ('contentlet', 'containers', 'template')");
+        dc.setMaxRows(1);
+        try {
+            final List<Map<String, Object>> results = dc.loadObjectResults();
+            final Date minDate = (Date) results.get(0).get("min_date");
+            if (null != minDate) {
+                startDate.setTime(minDate);
+            }
+        } catch (final DotDataException e) {
+            Logger.error(CMSMaintenanceFactory.class, String.format("An error occurred when finding the minimum date." +
+                    " Using date specified by user instead: '%s'", e.getMessage()), e);
+            startDate = null;

```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Assigning an Object to null is a code smell.  Consider refactoring.](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5534906)